File: android/app/src/main/java/app/foldgpt/install/InactiveIntegrationBundle.java
package app.foldgpt.install;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.*;

/** Authenticated, bounded release container. No filesystem writes or tar parser.
 * The caller supplies the independently trusted release SHA, never a digest
 * read from this container or calculated from an untrusted download. */
public final class InactiveIntegrationBundle {
    public static final String FORMAT="foldgpt.inactive-integration.v1";
    public static final String GPU_PREFIX="opt/foldgpt-gpu/mesa-26.2.2-foldgpt5";
    public static final String GPU_SHA="e02091631e5f16efbc3678373b2c048ebf81b10d551caf210d61b1954b7671d4";
    public static final String XKB="usr/share/X11/xkb";
    static final String CONTRACT_PATH="usr/local/share/foldgpt/launch-contract.v1";
    static final String CONTRACT="foldgpt.launch-contract.v1\n"
        +"scope=declared-launch-inputs-only\nguest=/usr/local/bin/foldgpt-session\nguest-shell=/bin/bash\n"
        +"guest-identity=etc/foldgpt-user-and-passwd-and-group\nguest-display=:2\nnative-xkb=usr/share/X11/xkb\n"
        +"gpu-prefix=/"+GPU_PREFIX+"\ngpu-driver=zink\nvulkan-icd=/"+GPU_PREFIX+"/share/vulkan/icd.d/freedreno_icd.aarch64.json\n"
        +"cdp-loopback=127.0.0.1:9223\nscale=android-display-density\nbridges=android-process-uid\n"
        +"android-root=not-required\nactivation=separate-validator-required\n";
    public static final int MAX_BYTES=64*1024*1024;
    static final Map<String,Integer> FILES;
    static final Map<String,String> LINKS;
    static {
        Map<String,Integer> files=new TreeMap<>();
        files.put("usr/local/bin/foldgpt-session",0700); files.put("usr/local/bin/xdg-open",0755);
        for(String name:List.of("foldgpt_keyring.py","foldgpt_ime.py","keyboard-focus.js")) files.put("usr/local/lib/foldgpt/"+name,0644);
        for(String name:List.of("initialize_keyring.py","supervise_keyring.py")) files.put("usr/local/lib/foldgpt/install/"+name,0600);
        files.put("usr/local/share/doc/foldgpt/LICENSE",0644); files.put(CONTRACT_PATH,0644);
        for(String name:List.of("lib/dri/libdril_dri.so","lib/libEGL.so.1.0.0","lib/libgallium-26.2.2.so","lib/libGL.so.1.2.0",
                "lib/libGLESv1_CM.so.1.1.0","lib/libGLESv2.so.2.0.0","lib/libvulkan_freedreno.so")) files.put(GPU_PREFIX+"/"+name,0755);
        for(String name:List.of("share/drirc.d/00-mesa-defaults.conf","share/drirc.d/00-turnip-defaults.conf","share/drirc.d/00-zink-defaults.conf",
                "share/vulkan/icd.d/freedreno_icd.aarch64.json")) files.put(GPU_PREFIX+"/"+name,0644);
        FILES=Collections.unmodifiableMap(files);
        Map<String,String> links=new TreeMap<>();
        links.put("lib/dri/zink_dri.so","libdril_dri.so"); links.put("lib/libEGL.so","libEGL.so.1"); links.put("lib/libEGL.so.1","libEGL.so.1.0.0");
        links.put("lib/libGL.so","libGL.so.1"); links.put("lib/libGL.so.1","libGL.so.1.2.0");
        links.put("lib/libGLESv1_CM.so","libGLESv1_CM.so.1"); links.put("lib/libGLESv1_CM.so.1","libGLESv1_CM.so.1.1.0");
        links.put("lib/libGLESv2.so","libGLESv2.so.2"); links.put("lib/libGLESv2.so.2","libGLESv2.so.2.0.0");
        Map<String,String> prefixed=new TreeMap<>(); links.forEach((path,target) -> prefixed.put(GPU_PREFIX+"/"+path,target));
        LINKS=Collections.unmodifiableMap(prefixed);
    }
    static final class Entry {
        final String scope,kind,path,sha,link; final int mode,size; int offset;
        Entry(String[] fields) throws IOException {
            scope=fields[0]; kind=fields[1]; path=path(fields[5]); sha=fields[4]; link=fields[6];
            if(!fields[2].matches("0[0-7]{3}") || !fields[3].matches("0|[1-9][0-9]{0,7}")) throw new IOException("Invalid integration size or mode");
            try { mode=Integer.parseInt(fields[2],8); size=Integer.parseInt(fields[3]); }
            catch(NumberFormatException error) { throw new IOException("Integration number overflow",error); }
            if(size>32*1024*1024 || !(scope.equals("I") || scope.equals("V"))) throw new IOException("Invalid integration scope or file bound");
            if(kind.equals("F")) { if(!sha.matches("[0-9a-f]{64}") || !link.equals("-")) throw new IOException("Invalid regular integration entry"); }
            else if(kind.equals("D") || kind.equals("L")) {
                if(size!=0 || !sha.equals("-") || kind.equals("D") && !link.equals("-")) throw new IOException("Invalid non-file integration entry");
                if(kind.equals("L") && (!path(link).equals(link) || link.contains("/"))) throw new IOException("Only same-directory integration links are accepted");
            } else throw new IOException("Unknown integration entry kind");
        }
    }
    final byte[] bytes;
    final NavigableMap<String,Entry> entries;
    public final String sha256,baseSha256,guestSha256,manifestSha256;
    private InactiveIntegrationBundle(byte[] data,String sha,String base,String guest,String manifest,NavigableMap<String,Entry> entries) {
        bytes=data; sha256=sha; baseSha256=base; guestSha256=guest; manifestSha256=manifest; this.entries=entries;
    }
    public static InactiveIntegrationBundle read(InputStream input,String trustedSha256,long trustedBytes) throws IOException {
        if(trustedSha256==null || !trustedSha256.matches("[0-9a-f]{64}") || trustedBytes<=0 || trustedBytes>MAX_BYTES)
            throw new IOException("Independent integration digest and size are required");
        byte[] bytes=input.readNBytes((int)trustedBytes+1);
        if(bytes.length!=trustedBytes || !hash(bytes).equals(trustedSha256)) throw new IOException("Integration container authentication failed");
        DataInputStream stream=new DataInputStream(new ByteArrayInputStream(bytes));
        byte[] magic=(FORMAT+"\n").getBytes(StandardCharsets.US_ASCII);
        if(!Arrays.equals(magic,stream.readNBytes(magic.length))) throw new IOException("Invalid integration framing");
        int length=stream.readInt();
        if(length<=0 || length>512*1024) throw new IOException("Integration manifest exceeds bound");
        byte[] encoded=stream.readNBytes(length);
        if(encoded.length!=length) throw new IOException("Truncated integration manifest");
        for(byte value:encoded) if(value<0 || value==0 || value=='\r') throw new IOException("Integration manifest must be canonical ASCII");
        String[] lines=new String(encoded,StandardCharsets.US_ASCII).split("\n",-1);
        if(lines.length<8 || !lines[0].equals(FORMAT) || !lines[lines.length-1].isEmpty()) throw new IOException("Invalid integration manifest header");
        String base=binding(lines[1],"base"),guest=binding(lines[2],"guest"),gpu=binding(lines[3],"gpu");
        if(!gpu.equals(GPU_SHA)) throw new IOException("Only the reviewed foldgpt5 GPU artifact is accepted");
        NavigableMap<String,Entry> entries=new TreeMap<>(); String previous="";
        Set<String> files=new TreeSet<>(),links=new TreeSet<>();
        int offset=magic.length+4+length;
        for(int index=4;index<lines.length-1;index++) {
            String[] fields=lines[index].split("\t",-1);
            if(fields.length!=7) throw new IOException("Invalid integration record width");
            Entry entry=new Entry(fields);
            if(entry.path.compareTo(previous)<=0) throw new IOException("Integration entries must be unique and sorted"); previous=entry.path;
            if(entry.scope.equals("I")) {
                if(entry.kind.equals("F")) {
                    if(!Objects.equals(FILES.get(entry.path),entry.mode) || entry.size==0) throw new IOException("Unexpected installed integration file or mode");
                    files.add(entry.path); entry.offset=offset;
                    if((long)offset+entry.size>bytes.length || !hash(bytes,offset,entry.size).equals(entry.sha)) throw new IOException("Integration payload hash or length differs");
                    offset+=entry.size;
                } else if(entry.kind.equals("L") && entry.mode==0777 && Objects.equals(LINKS.get(entry.path),entry.link)) links.add(entry.path);
                else throw new IOException("Unexpected installed integration link or directory");
            } else {
                if(!entry.path.equals(XKB) && !entry.path.startsWith(XKB+"/")) throw new IOException("Verification scope is restricted to Debian XKB");
                if(entry.mode!=(entry.kind.equals("D")?0755:entry.kind.equals("L")?0777:0644)) throw new IOException("Unexpected reviewed XKB mode");
            }
            entries.put(entry.path,entry);
        }
        if(!files.equals(FILES.keySet()) || !links.equals(LINKS.keySet()) || offset!=bytes.length) throw new IOException("Missing payload or trailing integration bytes");
        for(String path:List.of(XKB,XKB+"/rules/base",XKB+"/rules/evdev")) {
            Entry entry=entries.get(path);
            if(entry==null || !entry.scope.equals("V") || !entry.kind.equals(path.equals(XKB)?"D":"F")) throw new IOException("Required XKB record absent");
        }
        for(Entry entry:entries.values()) if(entry.scope.equals("V")) {
            if(!entry.path.equals(XKB)) {
                Entry parent=entries.get(entry.path.substring(0,entry.path.lastIndexOf('/')));
                if(parent==null || !parent.kind.equals("D")) throw new IOException("XKB manifest parent is not a directory");
            }
            if(entry.kind.equals("L")) {
                Entry target=entries.get(entry.path.substring(0,entry.path.lastIndexOf('/')+1)+entry.link);
                if(target==null || !target.kind.equals("F")) throw new IOException("XKB native link target is not an inventoried file");
            }
        }
        Entry contract=entries.get(CONTRACT_PATH);
        if(!hash(CONTRACT.getBytes(StandardCharsets.US_ASCII)).equals(contract.sha)) throw new IOException("Launch contract differs from runtime integration contract");
        return new InactiveIntegrationBundle(bytes,trustedSha256,base,guest,hash(encoded),entries);
    }
    static String path(String value) throws IOException {
        if(!value.matches("[A-Za-z0-9_+./-]{1,256}") || value.startsWith("/") || value.endsWith("/")) throw new IOException("Invalid integration path");
        for(String part:value.split("/",-1)) if(part.isEmpty() || part.equals(".") || part.equals("..")) throw new IOException("Noncanonical integration path");
        return value;
    }
    private static String binding(String line,String name) throws IOException {
        if(!line.matches(name+"\t[0-9a-f]{64}")) throw new IOException("Invalid integration input binding"); return line.substring(name.length()+1);
    }
    static String hash(byte[] value) throws IOException { return hash(value,0,value.length); }
    static String hash(byte[] value,int offset,int size) throws IOException {
        try { MessageDigest digest=MessageDigest.getInstance("SHA-256"); digest.update(value,offset,size); return hex(digest.digest()); }
        catch(java.security.NoSuchAlgorithmException error) { throw new IOException("SHA-256 unavailable",error); }
    }
    static String hex(byte[] data) { StringBuilder value=new StringBuilder(); for(byte item:data) value.append(String.format(Locale.ROOT,"%02x",item&255)); return value.toString(); }
}

File: tools/install/inactive_integration_bundle.py
"""Authenticate inputs and build the inactive native integration container.

This is a release assembly tool, not an Android installer. Debian XKB files are
inventoried for in-place verification; only FoldGPT scripts and the reviewed
Mesa runtime subset are included as new payload. No profile or device is read.
"""
import argparse
import hashlib
import io
import json
from pathlib import Path, PurePosixPath
import re
import struct
import tarfile

import guest_bundle

LEGACY_FORMAT = "foldgpt.inactive-integration.v1"
FORMAT = "foldgpt.inactive-integration.v2"
MAGIC = (FORMAT + "\n").encode("ascii")
GPU_PREFIX = "opt/foldgpt-gpu/mesa-26.2.2-foldgpt5"
GPU_SHA = "e02091631e5f16efbc3678373b2c048ebf81b10d551caf210d61b1954b7671d4"
BASE_SHA = "dd0aac2065057596d4210848eab198f3c3abd43dad2baa4622f5537e4ad3279f"
MAX_BUNDLE = 64 * 1024 * 1024
XKB = "usr/share/X11/xkb"
GPU_FILES = (
    "lib/dri/libdril_dri.so", "lib/libEGL.so.1.0.0", "lib/libgallium-26.2.2.so",
    "lib/libGL.so.1.2.0", "lib/libGLESv1_CM.so.1.1.0", "lib/libGLESv2.so.2.0.0",
    "lib/libvulkan_freedreno.so", "share/drirc.d/00-mesa-defaults.conf",
    "share/drirc.d/00-turnip-defaults.conf", "share/drirc.d/00-zink-defaults.conf",
    "share/vulkan/icd.d/freedreno_icd.aarch64.json",
)
GPU_LINKS = {
    "lib/dri/zink_dri.so": "libdril_dri.so", "lib/libEGL.so": "libEGL.so.1",
    "lib/libEGL.so.1": "libEGL.so.1.0.0", "lib/libGL.so": "libGL.so.1",
    "lib/libGL.so.1": "libGL.so.1.2.0", "lib/libGLESv1_CM.so": "libGLESv1_CM.so.1",
    "lib/libGLESv1_CM.so.1": "libGLESv1_CM.so.1.1.0", "lib/libGLESv2.so": "libGLESv2.so.2",
    "lib/libGLESv2.so.2": "libGLESv2.so.2.0.0",
}
LEGACY_CONTRACT_PATH = "usr/local/share/foldgpt/launch-contract.v1"
CONTEXT_HELPER = "usr/local/lib/foldgpt/foldgpt_agent_context.py"
CONTEXT_MANIFEST = "usr/local/share/foldgpt/agent-environment.v1.json"
CONTRACT_PATH = "usr/local/share/foldgpt/launch-contract.v2"
LEGACY_CONTRACT = ("foldgpt.launch-contract.v1\n"
            "scope=declared-launch-inputs-only\n"
            "guest=/usr/local/bin/foldgpt-session\n"
            "guest-shell=/bin/bash\n"
            "guest-identity=etc/foldgpt-user-and-passwd-and-group\n"
            "guest-display=:2\n"
            "native-xkb=usr/share/X11/xkb\n"
            "gpu-prefix=/" + GPU_PREFIX + "\n"
            "gpu-driver=zink\n"
            "vulkan-icd=/" + GPU_PREFIX + "/share/vulkan/icd.d/freedreno_icd.aarch64.json\n"
            "cdp-loopback=127.0.0.1:9223\n"
            "scale=android-display-density\n"
            "bridges=android-process-uid\n"
            "android-root=not-required\n"
            "activation=separate-validator-required\n").encode("ascii")
CONTRACT = LEGACY_CONTRACT.replace(b"foldgpt.launch-contract.v1\n", b"foldgpt.launch-contract.v2\n") + (
    "agent-context-generator=/" + CONTEXT_HELPER + "\nagent-context-manifest=/" + CONTEXT_MANIFEST + "\n"
    "agent-context-consumer=local-codex-global-agents-md\nagent-context-selection=first-nonempty-override-then-agents\n"
    "agent-context-delivery=separate-runtime-observation-required\n").encode("ascii")


def digest(data):
    return hashlib.sha256(data).hexdigest()


def safe_path(value):
    if (not re.fullmatch(r"[A-Za-z0-9_+./-]{1,256}", value)
            or str(PurePosixPath(value)) != value or value.startswith("/")
            or any(part in (".", "..") for part in value.split("/"))):
        raise ValueError("Noncanonical integration path")
    return value


def entry(scope, kind, path, mode, data=b"", link="-"):
    safe_path(path)
    if link != "-":
        safe_path(link)
        if "/" in link:
            raise ValueError("Only same-directory links are accepted")
    return {"scope": scope, "kind": kind, "mode": mode, "size": len(data),
            "sha256": digest(data) if kind == "F" else "-", "path": path, "link": link}


def snapshot(path, expected, limit):
    data = guest_bundle.read_bounded(path, limit)
    if digest(data) != expected:
        raise ValueError("Input archive SHA-256 differs: " + Path(path).name)
    return data


def inventory_xkb(base_data):
    result = {}
    with tarfile.open(fileobj=io.BytesIO(base_data), mode="r:gz") as archive:
        for item in archive:
            path = item.name.removeprefix("./").rstrip("/")
            if path != XKB and not path.startswith(XKB + "/"):
                continue
            safe_path(path)
            if path in result or item.size > 1024 * 1024:
                raise ValueError("Duplicate or oversized XKB entry")
            if item.isdir():
                record = entry("V", "D", path, item.mode)
            elif item.issym():
                record = entry("V", "L", path, item.mode, link=item.linkname)
            elif item.isfile():
                data = archive.extractfile(item).read(1024 * 1024 + 1)
                if len(data) != item.size:
                    raise ValueError("Truncated XKB file")
                record = entry("V", "F", path, item.mode, data)
            else:
                raise ValueError("Unsupported XKB entry type")
            if record["mode"] != {"D": 0o755, "F": 0o644, "L": 0o777}[record["kind"]]:
                raise ValueError("Unexpected reviewed XKB permissions")
            result[path] = record
    for required in (XKB, XKB + "/rules/base", XKB + "/rules/evdev"):
        if required not in result:
            raise ValueError("Required native XKB entry missing")
    for record in result.values():
        if record["kind"] == "L":
            target = str(PurePosixPath(record["path"]).parent / record["link"])
            if target not in result or result[target]["kind"] != "F":
                raise ValueError("Native XKB link does not resolve to an inventoried file")
    return result


def gpu_payload(gpu_data):
    selected = {GPU_PREFIX + "/" + path for path in (*GPU_FILES, *GPU_LINKS)}
    records, payload, seen = {}, {}, set()
    with tarfile.open(fileobj=io.BytesIO(gpu_data), mode="r:gz") as archive:
        for item in archive:
            path = item.name.removeprefix("./").rstrip("/")
            safe_path(path)
            if path in seen:
                raise ValueError("Duplicate Mesa archive path")
            seen.add(path)
            if path not in selected:
                continue
            relative = path[len(GPU_PREFIX) + 1:]
            if relative in GPU_LINKS:
                if not item.issym() or item.linkname != GPU_LINKS[relative]:
                    raise ValueError("Mesa runtime link contract differs")
                records[path] = entry("I", "L", path, 0o777, link=item.linkname)
            else:
                if not item.isfile() or item.size <= 0 or item.size > 32 * 1024 * 1024:
                    raise ValueError("Invalid Mesa runtime file")
                data = archive.extractfile(item).read(32 * 1024 * 1024 + 1)
                if len(data) != item.size:
                    raise ValueError("Truncated Mesa runtime file")
                if relative.startswith("lib/") and (data[:6] != b"\x7fELF\x02\x01" or data[18:20] != b"\xb7\x00"):
                    raise ValueError("Mesa runtime must be AArch64 ELF64 little-endian")
                mode = 0o755 if relative.startswith("lib/") else 0o644
                records[path] = entry("I", "F", path, mode, data)
                payload[path] = data
    if records.keys() != selected:
        raise ValueError("Reviewed Mesa runtime subset is incomplete")
    icd = json.loads(payload[GPU_PREFIX + "/share/vulkan/icd.d/freedreno_icd.aarch64.json"])
    if icd.get("ICD", {}).get("library_path") != "/" + GPU_PREFIX + "/lib/libvulkan_freedreno.so":
        raise ValueError("Mesa ICD does not select the reviewed prefix")
    return records, payload


def encode(records, payload, base_sha, guest_sha, gpu_sha):
    lines = [FORMAT, "base\t" + base_sha, "guest\t" + guest_sha, "gpu\t" + gpu_sha]
    for path, record in sorted(records.items()):
        lines.append("\t".join((record["scope"], record["kind"], format(record["mode"], "04o"),
                                str(record["size"]), record["sha256"], path, record["link"])))
    manifest = ("\n".join(lines) + "\n").encode("ascii")
    if len(manifest) > 512 * 1024:
        raise ValueError("Integration manifest exceeds bound")
    data = MAGIC + struct.pack(">I", len(manifest)) + manifest + b"".join(payload[path] for path in sorted(payload))
    if len(data) > MAX_BUNDLE:
        raise ValueError("Integration container exceeds bound")
    return data, manifest


def build(guest_path, guest_sha, gpu_path, base_path):
    sources = guest_bundle.verify(snapshot(guest_path, guest_sha, guest_bundle.MAX_BUNDLE_BYTES), guest_sha)
    records, payload = gpu_payload(snapshot(gpu_path, GPU_SHA, 16 * 1024 * 1024))
    records.update(inventory_xkb(snapshot(base_path, BASE_SHA, 400 * 1024 * 1024)))
    for path in guest_bundle.MODES:
        target = "usr/local/share/doc/foldgpt/LICENSE" if path == "LICENSE" else path.removeprefix("payload/")
        mode = guest_bundle.MODES[path]
        # These two are already installed 0600 by inactive keyring preparation.
        if target.startswith("usr/local/lib/foldgpt/install/"):
            mode = 0o600
        payload[target] = sources[path]
        records[target] = entry("I", "F", target, mode, sources[path])
    payload[CONTRACT_PATH] = CONTRACT
    records[CONTRACT_PATH] = entry("I", "F", CONTRACT_PATH, 0o644, CONTRACT)
    return encode(records, payload, BASE_SHA, guest_sha, GPU_SHA)


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--guest-bundle", type=Path, required=True)
    parser.add_argument("--guest-sha256", required=True)
    parser.add_argument("--gpu-archive", type=Path, required=True)
    parser.add_argument("--base-archive", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    data, manifest = build(args.guest_bundle, args.guest_sha256, args.gpu_archive, args.base_archive)
    guest_bundle.write_new_archive(args.output, data)
    guest_bundle.write_new_archive(args.output.with_suffix(".manifest"), manifest)
    print(json.dumps({"format": FORMAT, "bytes": len(data), "sha256": digest(data),
                      "manifestSha256": digest(manifest), "activated": False}))


if __name__ == "__main__":
    main()
